src/v/jl/jls_writer/jls_writer.cc

```diff
@@ -0,0 +1,157 @@
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "jl/jls_writer/jls_writer.h"
+
+#include "base/likely.h"
+#include "base/vassert.h"
+#include "bytes/iobuf.h"
+#include "bytes/iostream.h"
+#include "jl/jls/jls.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_types.h"
+#include "serde/rw/rw.h"
+
+#include <seastar/core/future.hh>
+
+#include <__utility/move.h>
+
+#include <stdexcept>
+#include <stdint.h>
+#include <utility>
+
+namespace {
+void write_ntp(iobuf& buf, const model::ntp& ntp) {
+    serde::write(buf, ntp.ns);
+    serde::write(buf, ntp.tp.topic);
+    serde::write(buf, ntp.tp.partition);
+}
+} // namespace
+
+namespace jl::jls_writer {
+
+ss::future<> writer::write(model::ntp ntp, model::record_batch&& batch) {
+    vassert(
+      batch.header().type == model::record_batch_type::raft_data,
+      "Only raft data expected in the jls writer");
+
+    if (unlikely(_closed)) {
+        throw std::runtime_error("writer is closed");
+    }
+
+    if (unlikely(!_did_write_header)) {
+        co_await write_header();
+    }
+
+    if (ntp != _last_ntp || _chunk_buffer.empty()) {
+        if (unlikely(ntp < _last_ntp)) {
+            throw std::runtime_error(
+              "writer api invariant violation: ntps are out of order");
+        }
+
+        maybe_flush_batch_buffer();
+
+        write_ntp(_chunk_buffer, ntp);
+
+        _last_ntp = ntp;
+        _last_offset = model::offset{};
+    }
+
+    if (unlikely(_last_offset > batch.base_offset())) {
+        throw std::runtime_error(
+          "writer api invariant violation: offsets are out of order");
+    }
+
+    serde::write(_batch_buffer, std::move(batch));
+    _last_offset = batch.last_offset();
+
+    if (
+      _batch_buffer.size_bytes() + _chunk_buffer.size_bytes()
+      >= _opts.chunk_size_target) {
+        maybe_flush_batch_buffer();
+        co_await maybe_flush_chunk_buffer();
+
+        // TODO(nv): Checksum?
+    }
+}
+
+ss::future<> writer::close() {
+    if (unlikely(_closed)) {
+        throw std::runtime_error("writer is already closed");
+    }
+
+    maybe_flush_batch_buffer();
+    co_await maybe_flush_chunk_buffer();
+    co_await write_footer();
+    co_await write_footer_epilogue();
+
+    _closed = true;
+}
+
+ss::future<> writer::write_header() {
+    co_await _out->write(
+      jl::jls::RP2_MAGIC_SIGNATURE.data(), jl::jls::RP2_MAGIC_SIGNATURE.size());
+    _did_write_header = true;
+}
+
+void writer::maybe_flush_batch_buffer() {
+    if (_batch_buffer.empty()) {
+        return;
+    }
+
+    auto prev_buf = std::exchange(_batch_buffer, {});
+
+    serde::write(
+      _chunk_buffer,
+      static_cast<jl::jls::chunk_data_batches_size_t>(prev_buf.size_bytes()));
+    _chunk_buffer.append(std::move(prev_buf));
+}
+
+ss::future<> writer::maybe_flush_chunk_buffer() {
```

Comment:
agree with @Lazin that it seems like we may want to be buffering at a higher level.